Implement JITless JS entry thunk to Wasm

danlliu opened the pull request:
#### e41732697b999464e80dc3fe1c95900c6c438888
<pre>
Implement JITless JS entry thunk to Wasm
<a href="https://bugs.webkit.org/show_bug.cgi?id=277266">https://bugs.webkit.org/show_bug.cgi?id=277266</a>
<a href="https://rdar.apple.com/113768974">rdar://113768974</a>

Reviewed by Yusuke Suzuki.

This patch builds on Justin&apos;s previous patch (<a href="https://commits.webkit.org/280485@main)">https://commits.webkit.org/280485@main)</a> to fully implement JITless entry to Wasm  code. This now allows us to run WebAssembly without enabling JIT. The lack of generated JIT thunks should significantly improve memory performance on programs with many call signatures.

Performance Testing: (N=10)
---------------------------

JetStream 2.0: 431.3 +/- 0.96 pt; up from 429.6 +/- 0.77 pt (0.40% significant progression)
Membuster 5: 1.357 GB +/- 3 MB; up from 1.354 GB +/- 3 MB (0.18% insignificant regression)

* JSTests/microbenchmarks/wasm-cc-int-to-int.js:
(wasm_instance.exports):
* JSTests/wasm/lowExecutableMemory/executable-memory-oom.js:
* JSTests/wasm/lowExecutableMemory/exports-oom.js:
(failCount.0.catch): Deleted.
(failCount.0.module.undefined.catch): Deleted.
* JSTests/wasm/noJIT/noJIT.js:
* JSTests/wasm/stress/cc-double-to-double.js: Copied from JSTests/wasm/stress/cc-int-to-int.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f64.result.f64.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f64.result.f64.f64.add.local.x.f64.const.42.func.export.string_appeared_here.param.x.f64.result.f64.f64.add.local.x.call.test.f64.const.1337.func.export.string_appeared_here.param.x.f64.result.f64.local.x.f64.const.98.call_indirect.t.type.sig_test.i32.const.0.f64.add.async test):
* JSTests/wasm/stress/cc-f32-kitchen-sink.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.result.f32.f32.add.local.x0.f32.add.local.x1.f32.add.local.x2.f32.add.local.x3.f32.add.local.x4.f32.add.local.x5.f32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.result.f32.f32.add.local.x0.f32.add.local.x1.f32.add.local.x2.f32.add.local.x3.f32.add.local.x4.f32.add.local.x5.f32.add.local.x6.f32.add.local.x7.f32.add.local.x8.f32.add.local.x9.f32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.result.f32.f32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.f32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-float-to-float.js: Copied from JSTests/wasm/stress/cc-int-to-int.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f32.result.f32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.f32.const.42.func.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.call.test.f32.const.1337.func.export.string_appeared_here.param.x.f32.result.f32.local.x.f32.const.98.call_indirect.t.type.sig_test.i32.const.0.f32.add.async test.instance.exports.float):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f32.result.f32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.f32.const.42.func.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.call.test.f32.const.1337.func.export.string_appeared_here.param.x.f32.result.f32.local.x.f32.const.98.call_indirect.t.type.sig_test.i32.const.0.f32.add.async test):
* JSTests/wasm/stress/cc-funcref.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.funcref.result.funcref.local.x.func.test2.export.string_appeared_here.param.x.funcref.result.funcref.result.funcref.local.x.local.x.async test):
* JSTests/wasm/stress/cc-i32-kitchen-sink-neg.js: Copied from JSTests/wasm/stress/cc-i32-kitchen-sink.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.i32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i32-kitchen-sink.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.i32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i64-kitchen-sink-neg.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.i64.add.local.x7.i64.add.local.x8.i64.add.local.x9.i64.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.i64.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i64-kitchen-sink.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.i64.add.local.x7.i64.add.local.x8.i64.add.local.x9.i64.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.i64.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-infinite-int-glitch.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.result.i32.local.x.local.x.func.test2.export.string_appeared_here.param.x.i32.result.i32.result.i32.result.i32.local.x.local.x.local.x.async test):
* JSTests/wasm/stress/cc-int-to-int-cross-module-with-exception.js:
(noInline.doTest.async test):
* JSTests/wasm/stress/cc-int-to-int-cross-module.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-int-to-int-jit-to-llint.js:
(noInline.iter.async test):
* JSTests/wasm/stress/cc-int-to-int-memory.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.async test):
* JSTests/wasm/stress/cc-int-to-int-no-jit.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-int-to-int-tail-call.js:
* JSTests/wasm/stress/cc-int-to-int.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-lotsofints-to-int.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.xa.i32.param.xb.i32.param.xc.i32.param.xd.i32.param.xe.i32.param.xf.i32.param.xg.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.xa.i32.add.local.xb.i32.add.local.xc.i32.add.local.xd.i32.add.local.xe.i32.add.local.xf.local.xg.async test):
* JSTests/wasm/stress/cc-so-i-heard-you-like-parameters.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.psel2.export.string_appeared_here.param.req1.i32.param.req2.i32.result.i32.result.i32.i32.eq.local.req1.i32.const.1.then.i32.const.1.else.result.i32.i32.eq.local.req2.i32.const.1.then.i32.const.2.else.i32.const.0):
(wat.async test):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/interpreter/CallFrame.h:
* Source/JavaScriptCore/interpreter/CallFrameInlines.h:
(JSC::CallFrame::setCallee):
* Source/JavaScriptCore/jit/FPRInfo.h:
(JSC::FPRInfo::toArgumentRegister):
* Source/JavaScriptCore/jit/GPRInfo.h:
(JSC::GPRInfo::toArgumentRegister):
* Source/JavaScriptCore/llint/InPlaceInterpreter.cpp:
(JSC::IPInt::initialize):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::generateThunkWithJumpTo):
(JSC::LLInt::generateThunkWithJumpToPrologue):
(JSC::LLInt::generateThunkWithJumpToLLIntReturnPoint):
(JSC::LLInt::returnLocationThunk):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllJITOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::addCodeOrigin):
(JSC::Wasm::OptimizingJITCallee::getOrigin const):
(JSC::Wasm::JSEntrypointInterpreterCallee::JSEntrypointInterpreterCallee):
(JSC::Wasm::JSEntrypointInterpreterCallee::entrypointImpl const):
(JSC::Wasm::JSEntrypointInterpreterCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmCallee.h:
(JSC::Wasm::jsEntrypointMetadataForGPR): Deleted.
(JSC::Wasm::jsEntrypointMetadataForFPR): Deleted.
(JSC::Wasm::dumpJSEntrypointInterpreterCalleeMetadata): Deleted.
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
(JSC::Wasm::LLIntPlan::ensureEntrypoint):
(JSC::Wasm::LLIntPlan::tryCreateInterpretedJSToWasmCallee):
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:
(JSC::Wasm::OMGPlan::work):
* Source/JavaScriptCore/wasm/WasmOMGPlan.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::gprToIndex):
(JSC::Wasm::fprToIndex):
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmJITInterpreterCrashForSIMDParameters):
(JSC::Wasm::createJSToWasmJITInterpreter):
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/JSToWasm.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::m_jsToWasmInterpreterCallee):
(JSC::m_jsToWasmBoxedInterpreterCallee): Deleted.
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:

Canonical link: <a href="https://commits.webkit.org/281726@main">https://commits.webkit.org/281726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be34954d6e9ad51d43e47377f233cb5c45ce1a40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13350 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64705 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11321 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11596 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49154 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7862 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62808 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37376 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52654 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29986 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34063 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9880 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10234 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53876 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66434 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60021 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4718 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10009 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56521 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56703 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3918 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81776 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9143 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35938 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14238 "Found 1454 new JSC stress test failures: microbenchmarks/memcpy-wasm-large.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-medium.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-small.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm.js.no-cjit-validate-phases, microbenchmarks/wasm-cc-int-to-int.js.default-wasm, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.default-wasm, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-bbq, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-collect-continuously, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-eager-jettison, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-no-cjit ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37020 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38113 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on JSTests/microbenchmarks/wasm-cc-int-to-int.js:
```diff
@@ -16,7 +15,7 @@ var wasm_module = new WebAssembly.Module(wasm_code);
 var wasm_instance = new WebAssembly.Instance(wasm_module);
 const { test, test_with_call, test_with_call_indirect } = wasm_instance.exports
 
-for (let i = 0; i < 100000; ++i) {
```
Was this change intended?

Achierius commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -2778,6 +2778,10 @@ op(wasm_function_prologue, macro ()
     crash()
 end)
 
+op(wasm_function_prologue_simd_trampoline, macro ()
```
For my own elucidation -- how do you know whether to call this trampoline vs. the normal one when you're calling a wasm function? Is there a distinction between wasm functions which use any simd features vs. those which use none?

danlliu replied:
We call the correct trampoline based on if the function uses SIMD, which is set in the FunctionParser. When we set up the Callee object, we'll initialize it with the right prologue.

Achierius commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -630,7 +630,8 @@ static inline void disableAllJITOptions()
     Options::useConcurrentJIT() = false;
 
-    if (!OptionsHelper::wasOverridden(Options::useWebAssemblyID))
+    // if (!OptionsHelper::wasOverridden(Options::useWebAssemblyID))
```
Stray comment

hyjorc1 replied:
Let's remove this.

hyjorc1 commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -272,7 +475,7 @@ inline bool shouldJIT(unsigned functionIndex)
 }
 
-JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationWasmTriggerTierUpNow, void, (Instance* instance, uint32_t functionIndex))
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationWasmTriggerTierUpNow, void, (Instance * instance, uint32_t functionIndex))
```
Any reason to add a space between `Instance` and `*`?

danlliu replied:
I think this was a side effect of autoformatting; I'll go ahead and revert these random changes.

Achierius commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -38,21 +40,21 @@
-DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(IPIntCallee);
-DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(LLIntCallee);
+namespace JSC {
+namespace Wasm {
```
nit: this could just be namespace `JSC::Wasm`

hyjorc1 commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -313,7 +516,7 @@ JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationWasmTriggerTierUpNow, void, (Instance
 
 #if ENABLE(WEBASSEMBLY_OMGJIT)
-JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationWasmTriggerOSREntryNow, void, (Probe::Context& context))
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationWasmTriggerOSREntryNow, void, (Probe::Context & context))
```
Ditto. And for all `Context &` and `Instance *` in the following.

Achierius commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -413,22 +415,28 @@ const StackMap& OptimizingJITCallee::stackmap(CallSiteIndex callSiteIndex) const
+        Options::useJIT() ? (usesSIMD ? LLInt::wasmFunctionEntryThunkSIMD().code().retagged<WasmEntryPtrTag>()
+            : LLInt::wasmFunctionEntryThunk().code().retagged<WasmEntryPtrTag>()
+            ) : (usesSIMD ? CodePtr<CFunctionPtrTag>(LLInt::getCodeFunctionPtr<CFunctionPtrTag>(wasm_function_prologue_simd_trampoline)).retagged<WasmEntryPtrTag>()
+                : CodePtr<CFunctionPtrTag>(LLInt::getCodeFunctionPtr<CFunctionPtrTag>(wasm_function_prologue_trampoline)).retagged<WasmEntryPtrTag>())
```
Can this logic be broken out into a separate function? It's hard to parse nested ternary operators

Achierius commented on JSTests/wasm/stress/cc-int-to-int-cross-module.js:
```diff
@@ -39,7 +37,7 @@ async function test() {
     const { test_with_call, test_with_call_indirect } = instance2.exports
 
-    for (let i = 0; i < 10000000; ++i) {
+    for (let i = 0; i < 1000; ++i) {
```
If I understand correctly, you've changed a good number of the tests which previously used `@requireOptions("--useJITCage=0")` as a workaround for using too much memory on the runner to instead run with fewer iterations.

Two questions:
1. Could this be broken out into its own patch? It seems relatively atomic
2. Does this preserve the nature of the tests -- i.e. do we still hit all the JIT tiers that we expect to, or could the reduced iteration count (in some cases down to 50) prevent that?

danlliu replied:
Ah, these changes were from Justin's original patch. I'm not 100% sure the original reasoning behind reducing the iteration count, but I don't think it hurts to run it the original number of times.

kmiller68 replied:
I think we try to use 10k since that's usually sufficient to test everything (at least with the eager configurations). Anything more just ends up making tests take forever.

danlliu replied:
Updated tests to all run 10k to be consistent

justinmichaud replied:
Yeah, I lowered the iteration count so that these tests wouldn't break the infrastructure. 10k seems reasonable.

Achierius commented on JSTests/wasm/stress/cc-int-to-int-tail-call.js:
```diff
@@ -1,4 +1,4 @@
-//@ skip
+//@skip
```
Accident?

Constellation replied:
I think tail-call is not implemented correctly yet, so we should keep it skip.

justinmichaud replied:
Unrelated: Did Keith's follow-up fix fix the tail call OMG tests?

justinmichaud replied:
Ostensibly tail calls should work now

Achierius commented on Source/JavaScriptCore/jit/GPRInfo.h:
```diff
@@ -741,10 +779,10 @@ class GPRInfo {
     }
 
-    static GPRReg toArgumentRegister(unsigned index)
+    static constexpr GPRReg toArgumentRegister(unsigned index)
```
One other set of changes that might've been a good candidate for landing separately is the migration of various functions to use constexpr. Not saying you need to split it off but in the future doing so will make it easier to review (less ambiguity about what constitutes the "meat" of the patch)

Achierius commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -578,7 +578,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, webAssemblyIPIntTiersUpToBBQ, true, Normal, "Allow IPInt to tier up to BBQ"_s) \
     v(Bool, webAssemblyIPIntTiersUpToOMG, true, Normal, "Allow IPInt to tier up to OMG"_s) \
-    v(Bool, useInterpretedJSEntryWrappers, false, Normal, "Allow some JS->wasm wrappers to be replaced by jit-less versions."_s) \
+    v(Bool, useInterpretedJSEntryWrappers, true, Normal, "Allow some JS->wasm wrappers to be replaced by jit-less versions."_s) \
```
Again for my own understanding: why just some?

danlliu replied:
I forgot to update the documentation! Should be all now!

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -25,6 +25,10 @@
 
 #include "config.h"
 #include "JSToWasm.h"
+#include "bmalloc/Gigacage.h"
```
Use `<wtf/Gigacage.h>`.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -25,6 +25,10 @@
 #include "JSToWasm.h"
+#include "bmalloc/Gigacage.h"
+#include "wasm/WasmCallee.h"
+#include <bmalloc/GigacageConfig.h>
```
You should avoid including this since bmalloc may not be enabled.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -25,6 +25,10 @@
 #include "config.h"
 #include "JSToWasm.h"
+#include "bmalloc/Gigacage.h"
+#include "wasm/WasmCallee.h"
```
`wasm/` is not necessary.

justinmichaud replied:
@danlliu Do you use clangd? You should pass `-header-insertion=never` to your clangd plugin

danlliu replied:
Yup, I set that up last time but forgot to set it up on my new machine; I've updated that now.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -25,6 +25,8 @@
 #include "config.h"
 #include "WasmOperations.h"
+#include "bytecode/ObjectAllocationProfile.h"
+#include "heap/DeferGC.h"
```
`bytecode/` and `heap/` is not necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -469,6 +478,7 @@ void OptimizingJITCallee::linkExceptionHandlers(Vector<UnlinkedHandlerInfo> unli
 
-} } // namespace JSC::Wasm
+}
+} // namespace JSC::Wasm
```
Keep the original code.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -25,10 +25,12 @@
 
 #include "config.h"
 #include "WasmCallee.h"
+#include "llint/LLIntThunks.h"
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -364,8 +366,8 @@ void OptimizingJITCallee::addCodeOrigin(unsigned firstInlineCSI, unsigned lastIn
-    // The inline frame list is stored in postorder. For example:
-    // A { B() C() D { E() } F() } -> B C E D F A
+        // The inline frame list is stored in postorder. For example:
+        // A { B() C() D { E() } F() } -> B C E D F A
```
This change isn't necessary.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -279,7 +572,7 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, JSEnt
             wasmCallingConvention().prologueScratchGPRs[2],
 #endif
-            wasmCallingConvention().prologueScratchGPRs[1]
+                wasmCallingConvention().prologueScratchGPRs[1]
```
This change isn not necessary.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -262,7 +555,7 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, JSEnt
     // https://webassembly.github.io/spec/js-api/index.html#exported-function-exotic-objects
     // If parameters or results contain v128, throw a TypeError.
-    // Note: the above error is thrown each time the [[Call]] method is invoked.
+    // Note: the above error is thrown each time the [[Call]] method is invoked
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -362,6 +655,7 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, JSEnt
 
-} } // namespace JSC::Wasm
+}
+} // namespace JSC::Wasm
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.load64(CCallHelpers::Address { GPRInfo::regWA2, 2*8 }, CCallHelpers::stackPointerRegister);
+        jit.load64(CCallHelpers::Address { GPRInfo::regWA3, 3*8 }, CCallHelpers::stackPointerRegister);
+        jit.load64(CCallHelpers::Address { GPRInfo::regWA4, 4*8 }, CCallHelpers::stackPointerRegister);
+        jit.load64(CCallHelpers::Address { GPRInfo::regWA5, 5*8 }, CCallHelpers::stackPointerRegister);
```
Place spaces between `0` and `*` and `8` and ditto for the others.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(4 * 8), GPRInfo::regWA4, GPRInfo::regWA5);
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(6 * 8), GPRInfo::regWA6, GPRInfo::regWA7);
+#elif USE(JSVALUE64)
+        jit.load64(CCallHelpers::Address { GPRInfo::regWA0, 0*8 }, CCallHelpers::stackPointerRegister);
```
Also, isn't it the reversed one? It is loading a value to stackPointer.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(0 * 8), GPRInfo::regWA0, GPRInfo::regWA1);
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(2 * 8), GPRInfo::regWA2, GPRInfo::regWA3);
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(4 * 8), GPRInfo::regWA4, GPRInfo::regWA5);
+        jit.loadPair64(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(6 * 8), GPRInfo::regWA6, GPRInfo::regWA7);
```
Let's use `Address(CCallHelpers::stackPointerRegister, 0 * 8)` instead.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        // Pop argument space values
+        jit.addPtr(MacroAssembler::TrustedImmPtr(Wasm::JSEntrypointInterpreterCallee::RegisterStackSpaceAligned), CCallHelpers::stackPointerRegister);
+
+        for (int32_t i = 0; i < 30; ++i)
+            jit.storePtr(MacroAssembler::TrustedImmPtr(0xbeef), CCallHelpers::Address { CCallHelpers::stackPointerRegister, -i * static_cast<int32_t>(sizeof(Register)) });
```
Let's not do this when it is release build.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+
+        // Save return registers
+#if CPU(ARM64) || CPU(ARM64E)
+        jit.storePair64(GPRInfo::regWA0, GPRInfo::regWA1, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 0 * 8 });
+        jit.storePair64(GPRInfo::regWA2, GPRInfo::regWA3, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 2 * 8 });
+        jit.storePair64(GPRInfo::regWA4, GPRInfo::regWA5, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 4 * 8 });
+        jit.storePair64(GPRInfo::regWA6, GPRInfo::regWA7, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 6 * 8 });
```
You can also use this for X64 since they also have `storePair64` definition.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.loadPair32(CCallHelpers::stackPointerRegister, MacroAssembler::TrustedImm32(1*8), GPRInfo::regWA2, GPRInfo::regWA3);
+#endif
+
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

danlliu replied:
macOS X86 build breaks when I use `loadDouble64`; it's not defined for that architecture.

(edit, wrong function, and might be for the wrong comment location then)

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+#endif
+        }
+
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+
+        // Memory
+        {
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        // saveJSEntrypointInterpreterRegisters
+        jit.subPtr(CCallHelpers::TrustedImmPtr(Wasm::JSEntrypointInterpreterCallee::SpillStackSpaceAligned), CCallHelpers::stackPointerRegister);
+
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.storePair32(GPRInfo::regWA0, GPRInfo::regWA1, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 0 * 8 });
+        jit.storePair32(GPRInfo::regWA2, GPRInfo::regWA3, CCallHelpers::Address { CCallHelpers::stackPointerRegister, 1 * 8 });
+#endif
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        jit.callOperation<OperationPtrTag>(operationJSToWasmEntryWrapperBuildReturnFrame);
+
+        // Load return registers
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+
+        // restoreJSEntrypointInterpreterRegisters
+
+#if CPU(ARM64) || CPU(ARM64E)
```
You can also use this code for CPU(X86_64).

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+    };
+
+    for (unsigned i = 0; i < functionSignature.argumentCount(); ++i) {
+        RELEASE_ASSERT(jsFrameConvention.params[i].location.isStack());
```
This is hot so let's not use RELEASE_ASSERT.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationJSToWasmEntryWrapperBuildFrame, void, (void* sp, CallFrame* cfr))
+{
+    RELEASE_ASSERT(Options::useInterpretedJSEntryWrappers());
```
Let's not use it.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationJSToWasmEntryWrapperBuildReturnFrame, void, (void* sp, CallFrame* cfr))
+{
+    RELEASE_ASSERT(Options::useInterpretedJSEntryWrappers());
```
Let's not use it.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+        if (functionSignature.returnType(0).isI32())
+            result = JSValue::encode(jsNumber(*access.operator()<int32_t>(registerSpace, 0)));
+        else if (functionSignature.returnType(0).isI64())
+            result = JSValue::encode(JSBigInt::makeHeapBigIntOrBigInt32(instance->globalObject(), *access.operator()<int64_t>(registerSpace, 0)));
```
This can potentially throw OOM error. So JS error check is necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+        }
+        ObjectInitializationScope initializationScope(vm);
+        DeferGCForAWhile deferGCForAWhile(vm);
+        JSArray* resultArray = JSArray::tryCreateUninitializedRestricted(initializationScope, nullptr, instance->globalObject()->arrayStructureForIndexingTypeDuringAllocation(indexingType), functionSignature.returnCount());
```
This can potentially throw OOM error. So JS error check is necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+                    result = jsNumber(*access.operator()<int32_t>(registerSpace, gprToIndex(loc.jsr().payloadGPR()) * bytesForWidth(Width::Width64)));
+                    break;
+                case TypeKind::I64:
+                    result = JSBigInt::makeHeapBigIntOrBigInt32(instance->globalObject(), *access.operator()<int64_t>(registerSpace, gprToIndex(loc.jsr().payloadGPR()) * bytesForWidth(Width::Width64)));
```
This can potentially throw OOM error. So JS error check is necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -62,13 +64,214 @@
+                    result = jsNumber(*access.operator()<int32_t>(cfr, calleeSPOffsetFromFP + loc.offsetFromSP()));
+                    break;
+                case TypeKind::I64:
+                    result = JSBigInt::makeHeapBigIntOrBigInt32(instance->globalObject(), *access.operator()<int64_t>(cfr, calleeSPOffsetFromFP + loc.offsetFromSP()));
```
This can potentially throw OOM error. So JS error check is necessary.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        // Prepare frame
+        jit.setupArguments<decltype(operationJSToWasmEntryWrapperBuildReturnFrame)>(CCallHelpers::stackPointerRegister, GPRInfo::callFrameRegister);
+        jit.callOperation<OperationPtrTag>(operationJSToWasmEntryWrapperBuildReturnFrame);
+
```
This can throw an error. So we need to handle JS exception handling here.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -214,6 +218,294 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        }
+
+        // Prepare frame
+        jit.setupArguments<decltype(operationJSToWasmEntryWrapperBuildFrame)>(CCallHelpers::stackPointerRegister, GPRInfo::callFrameRegister);
+        jit.callOperation<OperationPtrTag>(operationJSToWasmEntryWrapperBuildFrame);
```
This can throw an error. So JS exception handling is necessary.

kmiller68 replied:
I don't think it does throw actually.

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -791,25 +790,25 @@ MacroAssemblerCodeRef<JSEntryPtrTag> returnLocationThunk(OpcodeID opcodeID, Opco
+        LLINT_RETURN_LOCATION(op_in_by_val)
+        LLINT_RETURN_LOCATION(op_enumerator_get_by_val)
+        LLINT_RETURN_LOCATION(op_enumerator_put_by_val)
+        LLINT_RETURN_LOCATION(op_enumerator_in_by_val)
```
Let's remove this change.

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -225,7 +225,6 @@ MacroAssemblerCodeRef<JITThunkPtrTag> wasmFunctionEntryThunkSIMD()
     return codeRef;
 }
 
-
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -92,7 +96,7 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
             wasmCallingConvention().prologueScratchGPRs[2],
 #endif
-            wasmCallingConvention().prologueScratchGPRs[1]
+                wasmCallingConvention().prologueScratchGPRs[1]
```
Let's undo this change.

Constellation reviewed (commented):
Discussed with Daniel. Still exception handling in JIT code side is necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -799,13 +1010,14 @@ JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationAllocateResultsArray, JSArray*, (JSWe
     VM& vm = instance->vm();
```
Let's change `operationAllocateResultsArray` to use JSC_DEFINE_JIT_OPERATION, and add exception handling to the code calling this function.

Constellation reviewed (approved):
r=me

danlliu committed:
Implement JITless JS entry thunk to Wasm
https://bugs.webkit.org/show_bug.cgi?id=277266
rdar://113768974

Reviewed by Yusuke Suzuki.

This patch builds on Justin's previous patch (https://commits.webkit.org/280485@main) to fully implement JITless entry to Wasm  code. This now allows us to run WebAssembly without enabling JIT. The lack of generated JIT thunks should significantly improve memory performance on programs with many call signatures.

Performance Testing: (N=10)
---------------------------

JetStream 2.0: 431.3 +/- 0.96 pt; up from 429.6 +/- 0.77 pt (0.40% significant progression)
Membuster 5: 1.357 GB +/- 3 MB; up from 1.354 GB +/- 3 MB (0.18% insignificant regression)

* JSTests/microbenchmarks/wasm-cc-int-to-int.js:
(wasm_instance.exports):
* JSTests/wasm/lowExecutableMemory/executable-memory-oom.js:
* JSTests/wasm/lowExecutableMemory/exports-oom.js:
(failCount.0.catch): Deleted.
(failCount.0.module.undefined.catch): Deleted.
* JSTests/wasm/noJIT/noJIT.js:
* JSTests/wasm/stress/cc-double-to-double.js: Copied from JSTests/wasm/stress/cc-int-to-int.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f64.result.f64.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f64.result.f64.f64.add.local.x.f64.const.42.func.export.string_appeared_here.param.x.f64.result.f64.f64.add.local.x.call.test.f64.const.1337.func.export.string_appeared_here.param.x.f64.result.f64.local.x.f64.const.98.call_indirect.t.type.sig_test.i32.const.0.f64.add.async test):
* JSTests/wasm/stress/cc-f32-kitchen-sink.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.result.f32.f32.add.local.x0.f32.add.local.x1.f32.add.local.x2.f32.add.local.x3.f32.add.local.x4.f32.add.local.x5.f32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.result.f32.f32.add.local.x0.f32.add.local.x1.f32.add.local.x2.f32.add.local.x3.f32.add.local.x4.f32.add.local.x5.f32.add.local.x6.f32.add.local.x7.f32.add.local.x8.f32.add.local.x9.f32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.result.f32.f32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.f32.param.x1.f32.param.x2.f32.param.x3.f32.param.x4.f32.param.x5.f32.param.x6.f32.param.x7.f32.param.x8.f32.param.x9.f32.param.x10.f32.param.x11.f32.f32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-float-to-float.js: Copied from JSTests/wasm/stress/cc-int-to-int.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f32.result.f32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.f32.const.42.func.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.call.test.f32.const.1337.func.export.string_appeared_here.param.x.f32.result.f32.local.x.f32.const.98.call_indirect.t.type.sig_test.i32.const.0.f32.add.async test.instance.exports.float):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.f32.result.f32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.f32.const.42.func.export.string_appeared_here.param.x.f32.result.f32.f32.add.local.x.call.test.f32.const.1337.func.export.string_appeared_here.param.x.f32.result.f32.local.x.f32.const.98.call_indirect.t.type.sig_test.i32.const.0.f32.add.async test):
* JSTests/wasm/stress/cc-funcref.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.funcref.result.funcref.local.x.func.test2.export.string_appeared_here.param.x.funcref.result.funcref.result.funcref.local.x.local.x.async test):
* JSTests/wasm/stress/cc-i32-kitchen-sink-neg.js: Copied from JSTests/wasm/stress/cc-i32-kitchen-sink.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.i32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i32-kitchen-sink.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.result.i32.i32.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.x10.i32.param.x11.i32.i32.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i64-kitchen-sink-neg.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.i64.add.local.x7.i64.add.local.x8.i64.add.local.x9.i64.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.i64.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-i64-kitchen-sink.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.f0.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.local.x7.func.f1.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.i64.add.local.x1.i64.add.local.x2.i64.add.local.x3.i64.add.local.x4.i64.add.local.x5.i64.add.local.x6.i64.add.local.x7.i64.add.local.x8.i64.add.local.x9.i64.add.local.x10.local.x11.func.f2.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.result.i64.i64.add.local.x0.local.x11.func.f3.export.string_appeared_here.param.x0.i64.param.x1.i64.param.x2.i64.param.x3.i64.param.x4.i64.param.x5.i64.param.x6.i64.param.x7.i64.param.x8.i64.param.x9.i64.param.x10.i64.param.x11.i64.i64.add.local.x0.local.x11.drop.async test):
* JSTests/wasm/stress/cc-infinite-int-glitch.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.result.i32.local.x.local.x.func.test2.export.string_appeared_here.param.x.i32.result.i32.result.i32.result.i32.local.x.local.x.local.x.async test):
* JSTests/wasm/stress/cc-int-to-int-cross-module-with-exception.js:
(noInline.doTest.async test):
* JSTests/wasm/stress/cc-int-to-int-cross-module.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-int-to-int-jit-to-llint.js:
(noInline.iter.async test):
* JSTests/wasm/stress/cc-int-to-int-memory.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.async test):
* JSTests/wasm/stress/cc-int-to-int-no-jit.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-int-to-int-tail-call.js:
* JSTests/wasm/stress/cc-int-to-int.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* JSTests/wasm/stress/cc-lotsofints-to-int.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x0.i32.param.x1.i32.param.x2.i32.param.x3.i32.param.x4.i32.param.x5.i32.param.x6.i32.param.x7.i32.param.x8.i32.param.x9.i32.param.xa.i32.param.xb.i32.param.xc.i32.param.xd.i32.param.xe.i32.param.xf.i32.param.xg.i32.result.i32.i32.add.local.x0.i32.add.local.x1.i32.add.local.x2.i32.add.local.x3.i32.add.local.x4.i32.add.local.x5.i32.add.local.x6.i32.add.local.x7.i32.add.local.x8.i32.add.local.x9.i32.add.local.xa.i32.add.local.xb.i32.add.local.xc.i32.add.local.xd.i32.add.local.xe.i32.add.local.xf.local.xg.async test):
* JSTests/wasm/stress/cc-so-i-heard-you-like-parameters.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.psel2.export.string_appeared_here.param.req1.i32.param.req2.i32.result.i32.result.i32.i32.eq.local.req1.i32.const.1.then.i32.const.1.else.result.i32.i32.eq.local.req2.i32.const.1.then.i32.const.2.else.i32.const.0):
(wat.async test):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/interpreter/CallFrame.h:
* Source/JavaScriptCore/interpreter/CallFrameInlines.h:
(JSC::CallFrame::setCallee):
* Source/JavaScriptCore/jit/FPRInfo.h:
(JSC::FPRInfo::toArgumentRegister):
* Source/JavaScriptCore/jit/GPRInfo.h:
(JSC::GPRInfo::toArgumentRegister):
* Source/JavaScriptCore/llint/InPlaceInterpreter.cpp:
(JSC::IPInt::initialize):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::generateThunkWithJumpTo):
(JSC::LLInt::generateThunkWithJumpToPrologue):
(JSC::LLInt::generateThunkWithJumpToLLIntReturnPoint):
(JSC::LLInt::returnLocationThunk):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllJITOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::addCodeOrigin):
(JSC::Wasm::OptimizingJITCallee::getOrigin const):
(JSC::Wasm::JSEntrypointInterpreterCallee::JSEntrypointInterpreterCallee):
(JSC::Wasm::JSEntrypointInterpreterCallee::entrypointImpl const):
(JSC::Wasm::JSEntrypointInterpreterCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmCallee.h:
(JSC::Wasm::jsEntrypointMetadataForGPR): Deleted.
(JSC::Wasm::jsEntrypointMetadataForFPR): Deleted.
(JSC::Wasm::dumpJSEntrypointInterpreterCalleeMetadata): Deleted.
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
(JSC::Wasm::LLIntPlan::ensureEntrypoint):
(JSC::Wasm::LLIntPlan::tryCreateInterpretedJSToWasmCallee):
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:
(JSC::Wasm::OMGPlan::work):
* Source/JavaScriptCore/wasm/WasmOMGPlan.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::gprToIndex):
(JSC::Wasm::fprToIndex):
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmJITInterpreterCrashForSIMDParameters):
(JSC::Wasm::createJSToWasmJITInterpreter):
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/JSToWasm.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::m_jsToWasmInterpreterCallee):
(JSC::m_jsToWasmBoxedInterpreterCallee): Deleted.
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:

Canonical link: https://commits.webkit.org/281726@main

iangrunert commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -624,344 +610,280 @@ end
-        elsif X86_64 or X86_64_WIN
-            # These must match the wasmToJS thunk, since the unwinder won't be able to tell the difference between us and them.
-            # See calleeSaveRegistersImpl
+        elsif X86_64
```
Should this apply to X86_64_WIN as well?

iangrunert commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -624,344 +610,280 @@ end
             loadpairq -2 * SlotSize[cfr], memoryBase, boundsCheckingSize
             loadp -3 * SlotSize[cfr], wasmInstance
-        elsif X86_64 or X86_64_WIN
+        elsif X86_64
```
Should this apply to X86_64_WIN as well?

iangrunert commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -624,344 +610,280 @@ end
+    # Memory
+    if ARM64 or ARM64E
+        loadpairq JSWebAssemblyInstance::m_cachedMemory[wasmInstance], memoryBase, boundsCheckingSize
+    elsif X86_64
```
Should this apply to X86_64_WIN as well?

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -207,13 +211,301 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+
+        // Memory
+        {
+#if CPU(ARM64) || CPU(ARM64E) || CPU(X86_64)
```
In C++ code (unlike offlineasm, confusingly) `CPU(ARM64)` includes `CPU(ARM64E)` so you don't need that case.

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -207,13 +211,301 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
         jit.setupArguments<decltype(operationAllocateResultsArray)>(GPRInfo::wasmContextInstancePointer, CCallHelpers::TrustedImmPtr(&typeDefinition), indexingType, savedResultsGPR);
         JIT_COMMENT(jit, "operationAllocateResultsArray");
         jit.callOperation<OperationPtrTag>(operationAllocateResultsArray);
+        exceptionChecks.append(jit.branchTestPtr(CCallHelpers::NonZero, GPRInfo::returnValueGPR2));
```
Since we don't have a VM readily available here it makes sense that we rely on the exception register but in the future I think it would be better to do 
```
static_assert(operationExceptionRegister<operationAllocateResultsArray>() != InvalidGPRReg, "We don't have a VM readily available so we rely on exception being returned");
exceptionChecks.append(jit.branchTestPtr(CCallHelpers::NonZero, operationExceptionRegister<operationAllocateResultsArray>()));
``` 
That way if the signature changes in the future this doesn't subtly fail, instead, it doesn't even compile.

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -207,13 +211,301 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
+        // Prepare frame
+        jit.setupArguments<decltype(operationJSToWasmEntryWrapperBuildReturnFrame)>(CCallHelpers::stackPointerRegister, GPRInfo::callFrameRegister);
+        jit.callOperation<OperationPtrTag>(operationJSToWasmEntryWrapperBuildReturnFrame);
+        exceptionChecks.append(jit.branchTestPtr(CCallHelpers::NonZero, GPRInfo::returnValueGPR2));
```
Ditto.
